funsor/delta.py

```diff
@@ -0,0 +1,109 @@
+from __future__ import absolute_import, division, print_function
+
+from collections import OrderedDict
+
+from six import add_metaclass
+
+import funsor.ops as ops
+from funsor.domains import reals
+from funsor.ops import Op
+from funsor.terms import Align, Binary, Funsor, FunsorMeta, Number, Variable, eager, to_funsor
+from funsor.torch import Tensor
+
+
+class DeltaMeta(FunsorMeta):
+    """
+    Wrapper to fill in defaults.
+    """
+    def __call__(cls, name, point, log_density=0):
+        point = to_funsor(point)
+        log_density = to_funsor(log_density)
+        return super(DeltaMeta, cls).__call__(name, point, log_density)
+
+
+@add_metaclass(DeltaMeta)
+class Delta(Funsor):
+    """
+    Normalized delta distribution binding a single variable.
+
+    :param str name: Name of the bound variable.
+    :param Funsor point: Value of the bound variable.
+    :param Funsor log_density: Optional log density to be added when evaluating
+        at a point. This is needed to make :class:`Delta` closed under
+        differentiable substitution.
+    """
+    def __init__(self, name, point, log_density=0):
+        assert isinstance(name, str)
+        assert isinstance(point, Funsor)
+        assert isinstance(log_density, Funsor)
+        assert log_density.output == reals()
+        inputs = OrderedDict([(name, point.output)])
+        inputs.update(point.inputs)
+        inputs.update(log_density.inputs)
+        output = reals()
+        super(Delta, self).__init__(inputs, output)
+        self.name = name
+        self.point = point
+        self.log_density = log_density
+
+    def eager_subs(self, subs):
+        value = None
+        index_part = []
+        for k, v in subs:
+            if k in self.inputs:
+                if k == self.name:
+                    value = v
+                else:
+                    assert self.name not in v.inputs
+                    index_part.append((k, v))
+        index_part = tuple(index_part)
+        if value is None and not index_part:
+            return self
+
+        name = self.name
+        point = self.point.eager_subs(index_part)
+        log_density = self.log_density.eager_subs(index_part)
+        if value is not None:
+            if isinstance(value, Variable):
+                name = value.name
+            elif isinstance(value, (Number, Tensor)) and isinstance(point, (Number, Tensor)):
+                return (value == point).all().log() + log_density
+            else:
+                # TODO Compute a jacobian, update log_prob, and emit another Delta.
+                raise ValueError('Cannot substitute a {} into a Delta'
+                                 .format(type(value).__name__))
```

Comment:
Would we always want to emit a `Delta` here?

Reply:
Hmm I haven't thought about this, but I guess we could emit `Delta` for bijective functions, or in case of a non-bijective function a mixture of `Deltas`. For example
```py
y = Variable('y', reals())
Delta('x',1)(x=y**2/2) == (Delta('y', -1/2) + Delta(y,+1/2)) / 2
```